### PR TITLE
[SPARK-5116][MLlib] Add extractor for SparseVector and DenseVector

### PR DESCRIPTION
**Pull request by coderxiang**
Add extractor for SparseVector and DenseVector in MLlib to save some code while performing pattern matching on Vectors. For example, previously we may use:

```
 vec match {
      case dv: DenseVector =>
        val values = dv.values
        ...
      case sv: SparseVector =>
        val indices = sv.indices
        val values = sv.values
        val size = sv.size
        ...
  }
```

with extractor it is:

```
vec match {
    case DenseVector(values) =>
      ...
    case SparseVector(size, indices, values) =>
      ...
}
```

### CODE REVIEW

**Comment by mengxr** [comment]
@coderxiang The change looks good to me. Do you plan to update the implementations with pattern matching?

**Comment by coderxiang** [comment]
@mengxr Do you mean update the existing implementations using the extractor? Sure I can do that.

**Comment by mengxr** [comment]
Yes, that would simply the code in several places.

File: mllib/src/main/scala/org/apache/spark/mllib/optimization/Gradient.scala [hunk 1]
```diff
@@ -64,11 +64,17 @@ class LogisticGradient extends Gradient {
     val gradientMultiplier = (1.0 / (1.0 + math.exp(margin))) - label
     val gradient = data.copy
     scal(gradientMultiplier, gradient)
+    val minusYP = if (label > 0) margin else -margin
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/optimization/Gradient.scala]: Could you rebase? This change doesn't belong to this PR.

**Comment by coderxiang** [comment]
@mengxr Guess I did an incorrect rebase before. Just updated thanks.

**Comment by mengxr** [comment]
LGTM pending Jenkins.
